src/hotspot/share/runtime/handshake.cpp

```diff
@@ -587,3 +597,124 @@ HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* ma
                        pr_ret == HandshakeState::_succeeded ? "including" : "excluding", p2i(match_op));
   return pr_ret;
 }
+
+void HandshakeState::lock() {
+  _lock.lock_without_safepoint_check();
+}
+
+void HandshakeState::unlock() {
+  _lock.unlock();
+}
+
+// The careful dance between thread suspension and exit is handled here.
+// The caller has notified the agents that we are exiting, before we go
+// on, we must check for a pending external suspend request and honor it
+// in order to not surprise the thread that made the suspend request.
+void HandshakeState::thread_exit() {
+  guarantee(JavaThread::current() == _handshakee, "Must be");
+  while (true) {
+    {
+      MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+      // If we have any safepoint or handshake pending deal with it else set exit.
+      if (!SafepointMechanism::should_process(_handshakee)) {
+        _handshakee->set_exiting();
+        return;
+      }
+    }
+    ThreadBlockInVM tbivm(_handshakee);
+    // We're done with this suspend request, but we have to loop around
+    // and check again. Eventually we will get the lock without a pending
+    // external suspend request and will be able to mark ourselves as
+    // exiting.
+  }
+}
```

Comment:
I need a little help learning the steps of this dance :)

We reach here in _thread_in_vm. We cannot be suspended in this state. There
might be another thread waiting to handshake us to suspend us but why can't we
just ignore that and do the `_handshakee->set_exiting();` even without locking
HandshakeState::_lock?

Adding a handshake operation is lock free, so even if the check
`SafepointMechanism::should_process(_handshakee)` in L619 returns false a new
operation to process could have been added concurrently such that
`SafepointMechanism::should_process(_handshakee)` would return true when we
execute `_handshakee->set_exiting();` in L620. What am I missing?

Reply:
`HandshakeState::suspend()` is a synchronous handshake and adding the
handshake to the queue is lock free, but the execution of the synchronous
handshake itself requires a `HandshakeState::claim_handshake()` call which
does acquire the lock in question. We (the suspend requester) hold the lock
while the handshake is being processed so we either detect that
_handshakee->set_exiting() won the race (in the target thread) or we (the
suspend requester) win the race of setting the suspend flag so the target
thread can't exit yet.

Hopefully that helps explain this dance.

Reply:
Hi Dan,

thanks for picking up my question!

> `HandshakeState::suspend()` is a synchronous handshake and adding the
> handshake to the queue is lock free, but the execution of the synchronous
> handshake itself requires a `HandshakeState::claim_handshake()` call which
> does acquire the lock in question.

My point would be that the attempt to execute the synchronous handshake for
suspending a thread that is just about to call HandshakeState::thread_exit()
cannot make progress (blocks) while the target thread is not safe
(_thread_in_vm).

A synchronous handshake requires the target thread to be in a safe state for the
requester to execute the handshake operation.  When executing
HandshakeState::thread_exit() the suspendee is _thread_in_vm. And the requester
will find it to be `_not_safe` when calling `possibly_can_process_handshake()`
before calling `HandshakeState::claim_handshake()` or when calling
`can_process_handshake()` afterwards. In both cases try_process() returns with
failure _not_safe and the lock is not held.

```c++
 546	  if (!possibly_can_process_handshake()) {
 547	    // JT is observed in an unsafe state, it must notice the handshake itself
 548	    return HandshakeState::_not_safe;
 549	  }
 550	
 551	  // Claim the mutex if there still an operation to be executed.
 552	  if (!claim_handshake()) {
 553	    return HandshakeState::_claim_failed;
 554	  }
 555	
 556	  // If we own the mutex at this point and while owning the mutex we
 557	  // can observe a safe state the thread cannot possibly continue without
 558	  // getting caught by the mutex.
 559	  if (!can_process_handshake()) {
 560	    _lock.unlock();
 561	    return HandshakeState::_not_safe;
 562	  }
```

So isn't being unsafe sufficient to sync with suspend requests?

Reply:
Interesting point that I didn't pick up from your previous comment.
Thanks for making it more clear for me. I need to mull on it for a bit.

Reply:
Technically you are correct.
I have tested to remove this is previously version and all tests passes fine.
The reason I kept it is because the suspender have identified a thread for suspension and deemed it suspendable, so we play nice.
To know why suspend failed the suspender must check if thread is exiting after a failed suspend. (originally I had a bug here which caused me to wrongfully introduce this from the beginning)
I'll remove it, since it simplifies the code and David's comments about this code is now out-of-line can be fixed.

Reply:
> 
> 
> Technically you are correct.
> I have tested to remove this is previously version and all tests passes fine.
> The reason I kept it is because the suspender have identified a thread for suspension and deemed it suspendable, so we play nice.
> To know why suspend failed the suspender must check if thread is exiting after a failed suspend. (originally I had a bug here which caused me to wrongfully introduce this from the beginning)

Thanks for the explanation. As I understand you want to be fault tolerant
towards somewhat sloppy agents. That's of course ok. It should be explained in a
comment though.

> I'll remove it, since it simplifies the code and David's comments about this code is now out-of-line can be fixed.

Also it does not eliminate the race as far as I can tell.

Reply:
> Also it does not eliminate the race as far as I can tell.

You are correct.